examples/experimental/native/flexible_primitives_mnist_pytorch/model_def.py

```diff
@@ -0,0 +1,199 @@
+        # )
+
+        self.lr_g = self.context._LRScheduler(
+            lr_scheduler=LambdaLR(self.opt_g, lr_lambda=lambda epoch: 0.95 ** epoch),
+            step_mode=LRScheduler.StepMode.STEP_EVERY_EPOCH,
+        )
+
+    def build_training_data_loader(self) -> DataLoader:
+        if not self.data_downloaded:
+            self.download_directory = data.download_dataset(
+                download_directory=self.download_directory,
+                data_config=self.context.get_data_config(),
+            )
+            self.data_downloaded = True
+
+        train_data = data.get_dataset(self.download_directory, train=True)
+        return DataLoader(train_data, batch_size=self.context.get_per_slot_batch_size())
+
+    def build_validation_data_loader(self) -> DataLoader:
+        if not self.data_downloaded:
+            self.download_directory = data.download_dataset(
+                download_directory=self.download_directory,
+                data_config=self.context.get_data_config(),
+            )
+            self.data_downloaded = True
+
+        validation_data = data.get_dataset(self.download_directory, train=False)
+        return DataLoader(validation_data, batch_size=self.context.get_per_slot_batch_size())
+
+    def train_batch(
+        self, batch: TorchData, model: nn.Module, epoch_idx: int, batch_idx: int
+    ) -> Dict[str, torch.Tensor]:
+        imgs, _ = batch
+
+        # train generator
+        # sample noise and match gpu device (or keep as cpu)
+        z = torch.randn(imgs.shape[0], self.context.get_hparam("latent_dim"))
+        z = self.context._to_device(z)
+
+        # generate images
+        generated_imgs = self.model.generator(z)
+
+        # log sampled images to current directory
+        sample_imgs = generated_imgs[:6]
+        grid = torchvision.utils.make_grid(sample_imgs)
+        self.logger.writer.add_image(f'generated_images_epoch_{epoch_idx}', grid, batch_idx)
+
+        # ground truth result (ie: all fake)
+        # put on GPU because we created this tensor inside training_loop
+        valid = torch.ones(imgs.size(0), 1)
+        valid = self.context._to_device(valid)
+
+        # adversarial loss is binary cross-entropy
+        g_loss = F.binary_cross_entropy(self.model.discriminator(generated_imgs), valid)
+
+        self.context._backward(g_loss)
+        self.context._step_optimizer(self.opt_g)
+
+
+        # train discriminator
+        # Measure discriminator's ability to classify real from generated samples
+
+        # how well can it label as real?
+        valid = torch.ones(imgs.size(0), 1)
+        valid = self.context._to_device(valid)
+
+        real_loss = F.binary_cross_entropy(self.model.discriminator(imgs), valid)
+
+        # how well can it label as fake?
+        fake = torch.zeros(imgs.size(0), 1)
+        fake = self.context._to_device(fake)
+
+        fake_loss = F.binary_cross_entropy(
+            self.model.discriminator(generated_imgs.detach()), fake)
```

Comment:
What is this `detatch()` call for?

Reply:
To avoid the gradients calculated by `backward(g_loss)` affect these calculated by `backward(d_loss).`

Reply:
oh cool.  TIL.